File: backend/livestream/domains/livestream.go
package domains

import (
	"context"
	"time"

	"github.com/gofrs/uuid/v5"
)

type LivestreamVisibility string

const (
	LivestreamPublicVisibility  LivestreamVisibility = "public"
	LivestreamPrivateVisibility                      = "private"
)

type Livestream struct {
	Id           uuid.UUID            `json:"id" db:"id"`
	UserId       uuid.UUID            `json:"userId" db:"user_id"`
	Title        string               `json:"title" db:"title"`
	Description  *string              `json:"description" db:"description"`
	ThumbnailURL *string              `json:"thumbnailUrl" db:"thumbnail_url"`
	ViewCount    int                  `json:"viewCount" db:"view_count"`
	Visibility   LivestreamVisibility `json:"visibility" db:"visibility"`
	StartedAt    time.Time            `json:"startedAt" db:"started_at"`
	EndedAt      *time.Time           `json:"endedAt" db:"ended_at"`
	CreatedAt    time.Time            `json:"createdAt" db:"created_at"`
	UpdatedAt    time.Time            `json:"updatedAt" db:"updated_at"`
	VODId        *uuid.UUID           `json:"vodId" db:"vod_id"`
}

type LivestreamRepository interface {
	GetById(ctx context.Context, id uuid.UUID) (*Livestream, error)
	GetByUser(ctx context.Context, userId uuid.UUID) (*Livestream, error)
	GetRecommendedLivestreams(ctx context.Context, page int, limit int) ([]Livestream, error)
	Create(ctx context.Context, ls Livestream) (*Livestream, error)
	Update(ctx context.Context, ls Livestream) (*Livestream, error)
	Delete(ctx context.Context, id uuid.UUID) error
	SetAuthorDisabled(ctx context.Context, userId uuid.UUID, disabled bool) error
	ReplaceDisabledAuthors(ctx context.Context, userIds []uuid.UUID) error
}

File: mobile/lib/models/livestream.dart
class Livestream {
  final String id;
  final String userId;
  final String title;
  final String? description;
  final String? thumbnailUrl;
  final int viewCount;
  final String visibility;
  final String startedAt;
  final String? endedAt;
  final String createdAt;
  final String updatedAt;
  final String? vodId;

  // Joined user info (from backend joined queries)
  final String? username;
  final String? displayName;
  final String? profilePicture;

  const Livestream({
    required this.id,
    required this.userId,
    required this.title,
    this.description,
    this.thumbnailUrl,
    this.viewCount = 0,
    this.visibility = 'public',
    required this.startedAt,
    this.endedAt,
    required this.createdAt,
    required this.updatedAt,
    this.vodId,
    this.username,
    this.displayName,
    this.profilePicture,
  });

  bool get isLive => endedAt == null;

  factory Livestream.fromJson(Map<String, dynamic> json) {
    return Livestream(
      id: json['id'] as String,
      userId: json['userId'] as String,
      title: json['title'] as String,
      description: json['description'] as String?,
      thumbnailUrl: json['thumbnailUrl'] as String?,
      viewCount: json['viewCount'] as int? ?? 0,
      visibility: json['visibility'] as String? ?? 'public',
      startedAt: json['startedAt'] as String,
      endedAt: json['endedAt'] as String?,
      createdAt: json['createdAt'] as String,
      updatedAt: json['updatedAt'] as String,
      vodId: json['vodId'] as String?,
      username: json['username'] as String?,
      displayName: json['displayName'] as String?,
      profilePicture: json['profilePicture'] as String?,
    );
  }
}
